staging/src/k8s.io/client-go/util/workqueue/queue_test.go

```diff
@@ -197,6 +197,72 @@ func TestReinsert(t *testing.T) {
 	}
 }
 
+func TestCollapse(t *testing.T) {
+	q := workqueue.New()
+	// Add a new one twice
+	q.Add("bar")
+	q.Add("bar")
+
+	// It should get the new one
+	i, _ := q.Get()
+	if i != "bar" {
+		t.Errorf("Expected %v, got %v", "bar", i)
+	}
+
+	// Finish that one up
+	q.Done(i)
+
+	// There should be no more objects in the queue
+	if a := q.Len(); a != 0 {
+		t.Errorf("Expected queue to be empty. Has %v items", a)
+	}
+}
+
+func TestCollapseWhileProcessing(t *testing.T) {
+	q := workqueue.New()
+	q.Add("foo")
+
+	// Start processing
+	i, _ := q.Get()
+	if i != "foo" {
+		t.Errorf("Expected %v, got %v", "foo", i)
+	}
+
+	// Add the same one twice
+	q.Add("foo")
+	q.Add("foo")
+
+	waitCh := make(chan struct{})
+	// simulate another worker consuming the queue
+	go func() {
+		defer close(waitCh)
+		i, _ := q.Get()
+		if i != "foo" {
+			t.Errorf("Expected %v, got %v", "foo", i)
+		}
+		// Finish that one up
+		q.Done(i)
+	}()
+
+	// give the worker some head start to avoid races
+	// on the select statement that cause flakiness
+	time.Sleep(100 * time.Millisecond)
```

Comment:
That can be flaky in heeavily starved ens, no?

Reply:
if is flaky it means we have a problem with the workqueue implementation, to ensure that the worker gets blocked waiting for the "foo" object to be "Done".

If we don't wait it can happen that the worker get the object without blocking and the test still pass

Reply:
OK - I guess the part that I was missing that even if we don't sleep (or during the sleep the goroutine will not start - which was the part that I was asking for), it will still pass because `Done()` will be called and only then the goroutine will fire and just work.

So we're good here.